CORE-6969 signature spec improvements

vlajos opened the pull request:
* CORE-6969 signature spec improvements
* CORE-6969 Phase out TransactionSignatureUtils in the favor of TransactionSignatureService

Follow up PR with integration test sandboxing tweaks: https://github.com/corda/corda-runtime-os/pull/2505

vlajos committed:
CORE-6969 Refactor TransactionSignatureUtils into TransactionSignatureService

vlajos committed:
CORE-6969 Use Signature verification through TransactionSignatureService

vlajos committed:
CORE-6969 Use the spec...

vlajos committed:
CORE-6969 Fix tests, detekt

vlajos committed:
CORE-6969 detekt

chrisr3 commented on components/ledger/ledger-utxo-flow/src/integrationTest/kotlin/net/corda/ledger/utxo/flow/impl/transaction/serializer/tests/UtxoSignedTransactionKryoSerializationTest.kt:
```diff
@@ -76,9 +61,15 @@ class UtxoSignedTransactionKryoSerializationTest {
+            jsonMarshallingService = setup.fetchService(1500)
+            transactionSignatureService = setup.fetchService(1500)
+            checkpointSerializerBuilderFactory = setup.fetchService(1500)
+
```
No, this should follow `WireTransactionKryoSerializationTest`.

chrisr3 commented on components/ledger/ledger-consensual-flow/src/integrationTest/kotlin/net/corda/ledger/consensual/flow/impl/transaction/serializer/tests/ConsensualSignedTransactionAMQPSerializationTest.kt:
```diff
@@ -91,18 +77,22 @@ class ConsensualSignedTransactionAMQPSerializationTest {
+            merkleTreeProvider = setup.fetchService(1500)
+            jsonMarshallingService = setup.fetchService(1500)
+            transactionSignatureService = setup.fetchService(1500)
+
```
These are the platform services. I suspect you want the sandbox services really.

chrisr3 commented on components/ledger/ledger-consensual-flow/src/integrationTest/kotlin/net/corda/ledger/consensual/flow/impl/transaction/serializer/tests/ConsensualSignedTransactionKryoSerializationTest.kt:
```diff
@@ -78,9 +61,15 @@ class ConsensualSignedTransactionKryoSerializationTest {
+            transactionSignatureService = setup.fetchService(1500)
+            checkpointSerializerBuilderFactory = setup.fetchService(1500)
+
             wireTransactionKryoSerializer = setup.fetchService(
```
No, see `WireTransactionKryoSerializationTest` for better approach.

chrisr3 commented on components/ledger/ledger-utxo-flow/src/integrationTest/kotlin/net/corda/ledger/utxo/flow/impl/transaction/serializer/tests/UtxoSignedTransactionAMQPSerializationTest.kt:
```diff
@@ -91,18 +77,22 @@ class UtxoSignedTransactionAMQPSerializationTest {
+            cipherSchemeMetadata = setup.fetchService(1500)
+            merkleTreeProvider = setup.fetchService(1500)
+            jsonMarshallingService = setup.fetchService(1500)
+            transactionSignatureService = setup.fetchService(1500)
```
These are the platform services. I suspect you want to use the sandbox services really.

kyriathar commented on components/ledger/ledger-common-flow-api/src/main/kotlin/net/corda/ledger/common/flow/transaction/TransactionSignatureService.kt:
```diff
@@ -0,0 +1,18 @@
+import net.corda.v5.crypto.SecureHash
+import java.security.PublicKey
+
+interface TransactionSignatureService {
+    @Suspendable
+    fun sign(transactionId: SecureHash, publicKey: PublicKey): DigitalSignatureAndMetadata
+
+    /**
+     * The underlying verification service signals the verification failures with different exceptions.
+     * [DigitalSignatureVerificationService]
+     */
+    fun verifySignature(transactionId: SecureHash, signatureWithMetadata: DigitalSignatureAndMetadata)
+}
```
I understand we want to provide a more convenient facility for signing/ verifying but not very sure these two APIs should be baked together. I had a look at Jira ticket and slack conversation, I think this should be a wrapper service around sign service and signature spec service, to eliminate `SignatureSpec` service parameter.

lankydan replied:
We could split the verification out to another service but i don't think it really matter since the service is internal anyway.

vlajos replied:
I thought it is cleaner to encapsulate the sigSpec saving/using and also the 2 places of SignableData creation.

lankydan commented on components/ledger/ledger-common-flow/build.gradle:
```diff
@@ -38,13 +45,28 @@ dependencies {
+    integrationTestImplementation project(':testing:sandboxes-testkit')
 
-    integrationTestRuntimeOnly project(':components:flow:flow-service')
+//    integrationTestRuntimeOnly project(':components:flow:flow-service')
```
Commented out stuff. Hoping this goes in so we can remove the flow dependency!!

vlajos replied:
The commented line is removed now.

chrisr3 reviewed (commented):
I updated `WireTransactionKryoSerializationTest` to create a FLOW sandbox using `FlowSandboxService`, and then use its `CheckpointSerializer` and `WireTransactionFactory` sandbox singleton services. (And also its `JsonMarshallingService`?!) I think this testing approach is more realistic than trying to construct everything "by hand" - and with the platform's services too.

lankydan commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt:
```diff
@@ -0,0 +1,99 @@
+
+@Suppress("Unused")
+@Component(
+    service = [TransactionSignatureService::class, UsedByFlow::class, UsedByPersistence::class, UsedByVerification::class],
```
`UsedByFlow` should be enough here.

vlajos replied:
Thanks. I have updated it.

kyriathar commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt:
```diff
@@ -0,0 +1,99 @@
+
+        val signedData = SignableData(transactionId, signatureWithMetadata.metadata)
+        digitalSignatureVerificationService.verify(
+            publicKey = signatureWithMetadata.by,
```
I think we should be injecting `MemberLookup` service (I believe) to check `signatureWithMetadata.by` exists in known counterparties, rather than just using returned signing key. `signatureWithMetadata.by` will be changing at some point to be public key hash instead to enforce not using the returned from wire signing key.

lankydan replied:
I think that makes sense.

vlajos replied:
We agreed with @kyriathar and @lankydan that we'll fix this issue when the hashed Publickeys will be available in the Signatures. I have created CORE-7993 for this.

vlajos commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt:
```diff
@@ -0,0 +1,99 @@
+
+@Suppress("Unused")
+@Component(
+    service = [TransactionSignatureService::class, UsedByFlow::class, UsedByPersistence::class, UsedByVerification::class],
```
I am not too sure if these are the correct sandboxes referenced here.

kyriathar reviewed (commented):
`SignatureSpecService` use and `SignatureSpec` being added to `DigitalSignatureAndMetadata` metadata and then being signed itself as well looks good to me.

lankydan commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt:
```diff
@@ -0,0 +1,99 @@
+
+    override fun verifySignature(transactionId: SecureHash, signatureWithMetadata: DigitalSignatureAndMetadata) {
+        val signatureSpecStr = signatureWithMetadata.metadata.properties[SIGNATURE_METADATA_SIGNATURE_SPEC_KEY]
+        require(signatureSpecStr != null) { "There are no signature spec in the Signature's metadata $signatureWithMetadata" }
```
Can use `requireNotNull`.

vlajos replied:
Thanks. I have updated it.

lankydan reviewed (commented):
Nice, think this PR simplifies the code quite a bit.

vlajos commented:
> I updated `WireTransactionKryoSerializationTest` to create a FLOW sandbox using `FlowSandboxService`, and then use its `CheckpointSerializer` and `WireTransactionFactory` sandbox singleton services. (And also its `JsonMarshallingService`?!) I think this testing approach is more realistic than trying to construct everything "by hand" - and with the platform's services too.

@chrisr3 Yes. Definitely. I created this one for the integration tests's tweaks: https://github.com/corda/corda-runtime-os/pull/2505
(It is essentially the last evening's tip of the current PR's branch, just I have split them to make the reviewers life a bit easier.)

vlajos committed:
CORE-6969 Remove not needed UsedBys

vlajos committed:
CORE-6969 PR review

vlajos committed:
CORE-6969 Remove duplicated services

vlajos committed:
CORE-6969 make TransactionSignatureService system service

vlajos committed:
CORE-6969 requireNotNull, Missing SignatureSpec.equal workaround

mcgovc reviewed (approved):
LGTM - build changes
